keps/sig-node/20191025-image-pull-progress.md

```diff
@@ -0,0 +1,314 @@
+like the overall amount of bytes already downloaded, the needed bytes to be
+downloaded and the estimated pull time based on the currently available
+bandwidth. These information are directly available via tools like `kubectl`,
+because each workload exposes this state directly to the Kubernetes API.
+
+## Motivation
+
+During Kubernetes workload creation, the time needed to pull the container image
+from the remote location varies depending on multiple factors. In low-bandwidth
+scenarios the image pull step might consume a significant amount time
+especially if the container image is large. Kubernetes provides currently no
+possibility to expose this pull progress to the user, whereas the only indicator
+is the `ContainerCreating` or `PullingImage` state.
+
+### Goals
+
+- Enhance the Container Runtime Interface (CRI) to be able to stream all
+  necessary information from the underlying container runtime
+- Expose the progress to the Kubernetes API
+- Providing a way how to display the progress to the end user via the CLI
+
+### Non-Goals
+
+- Everything which is not related to the container image pull procedure
+
+## Proposal / User Stories
+
+The overall implementation can be split up into three user stories.
+
+### CRI API enhancement
+
+The Kubernets runtime API `ImageService` has to be modified to provide an
+additional streaming RPC. This server-side stream provides continuously image
+pull related metadata during the whole image pull process. The new endpoint
+would be added in addition to the already available `PullImage` RPC and re-use
+the request data type:
+
+```protobuf
+service ImageService {
+    …
+
+    // PullImageStream pulls an image and provides continuous metadata during
+    // that time
+    rpc PullImageProgress(PullImageRequest) returns (stream PullImageProgressResponse) {}
+}
+
+enum PullImageState {
+    STARTED = 0;
+    PULLING = 1;
+    FAILED  = 2;
+    DONE    = 3;
+}
+
+message PullImageProgressResponse {
+    // The current state of the image pull proress
+    PullImageState state = 1;
+
+    // The overall size of the image in bytes
+    uint64 size = 2;
+
+    // The amount of data already retrieved in bytes
+    uint64 current_offset = 3;
+
+    // String indicating the reason for the current state, like a failure message
+    string reason = 4;
+}
+```
+
+The `PullImageState` provides richer information to the kubelet in which state
+the image pull currently resides as well as a fine granular error messages
+directly from the runtimes via the `reason` field.
+
+### Kubernetes API additions
+
+The already existing kubelet `imageManager` events `PullingImage`, `PulledImage`
+and `FailedToPullImage` will be re-used and extended with the additional
+metadata provided by the CRI. This means that the container will still reside in
+`ContainerStateWaiting` during the image pull, but the events regarding the
+image pull process will be thrown periodically.
```

Comment:
This has non-trivial scalability implications. I would like to have a better analysis at the KEP level.

In fact, that is a great candidate for pilot of scalability-related part of PRR:
https://github.com/kubernetes/community/pull/4184

@johnbelamaric - FYI

Reply:
I agree. I thought about having the same update period as the CRI stream updates are coming in. What about making it configurable? I think a value around 1-30 seconds would be reasonable depending on the bandwidth.

Reply:
Alright I will attach the answers in the first comment.

Reply:
I tried to answer all the questions from the questionnaire (see PR body), PTAL and bear with me If I misinterpreted something. :see_no_evil:

Reply:
I don't see any of those answered anywhere - can you please point me to them?
[BTW - it's a crazy week for me, so I may not get to it before end of week...]

Reply:
Yes, the first comment on this PR contains a drop down menu where the answered are listed :)

![Peek 2019-10-28 12-32](https://user-images.githubusercontent.com/695473/67675324-2517f280-f97f-11e9-86d0-72f89f627b53.gif)

Reply:
Aaah I see - now. Thanks. Will take a look.

Reply:
LOL @saschagrunert :)